### DISCUSSION TITLE
Questions about the `NumaNode` structure in `vmm` crate - Suggestions welcomed!

### DISCUSSION THREAD

**Discussion by MrXinWang** [cloud-hypervisor/cloud-hypervisor#2947] [Q&A] [answered]
Hi guys, I am trying to add the missing device tree support for NUMA on AArch64 platform. 

Background: The linux kernel has provides the device tree binding of NUMA on AArch64: https://www.kernel.org/doc/Documentation/devicetree/bindings/numa.txt

By going through the code, there is something I would like to discuss with the community first before I actually writing the code.

(1) Whether we want the AArch64 device tree NUMA support

For now, both of the AArch64 and x86_64 have ACPI support for NUMA. As the Cloud Hypervisor on AArch64 supports both device tree and ACPI, personally I would also like to add the device tree support because of following reasons:

- It does not make much sense to forbid users to use NUMA on AArch64 using `--numa` if they choose to use direct kernel boot.
- The device tree NUMA support code is not that hard to add...

(2) Where to locate the `NumaNodes` BTreemap and `NumaNode` structure.

The idea to add NUMA support for AArch64 device tree is that - We can pass the NUMA config - which I believe will eventually stored in the `NumaNodes` Btreemap, to the `arch::configure_system()` located in `arch/src/aarch64/mod.rs`. The `arch::configure_system()` will eventually call `create_fdt()`. We can then dynamically create cpu and memory nodes based on NUMA configuration using above device tree binding.

The `NumaNodes` Btreemap is currently in the `vmm` crate (`vmm/src/vm.rs`). However, for now, we cannot let the `arch` crate to import the `vmm` crate, as this will cause loop dependency. In that case, it seems that currently it is impossible to pass the `NumaNode` structure to the `arch` crate.

Proposals:

(1) I've had a discussion with @sboeuf and one of the possible solution is to move the `NumaNodes` BTreemap and `NumaNode` structure to the `arch` crate. If we do this:

- The import issue can be solved.
- We need to split the `NumaNodes` BTreemap and `NumaNode` structure to two duplicates, one for x86 and one for arm. Seems a little bit redundant.

(2) Add some parsing logic in the AArch64 `arch::configure_system()` to extract the information we want, and create a new BTreemap to make sure the mapping is the same. Then just pass these information, such as 

```rust
memory_regions: Vec<Arc<GuestRegionMmap>>,
hotplug_regions: Vec<Arc<GuestRegionMmap>>,
cpus: Vec<u8>,
distances: BTreeMap<u32, u8>,
memory_zones: Vec<String>,
```

to the `arch::configure_system()`. Since these fields are normal types, I think it should be fine. But to me the newly added logic is a little bit ugly. @michael2012z Your comments about this is welcome.

@rbradford @sboeuf @michael2012z What do you think?

**Comment by rbradford** [comment, marked as answer]
@MrXinWang Just move the structures to `arch/lib.rs` you don't need to do different ones for each architecture.

**Reply by MrXinWang** [reply]
Thanks @rbradford !
